dlib/cuda/cuda_dlib.cu

```diff
@@ -2242,6 +2242,131 @@ namespace dlib
                           dmeans.device(), dvars.device(), eps, src.num_samples(), num);
         }
 
+   // ----------------------------------------------------------------------------------------
+
+        __global__ void _cuda_rms_normalize(float* out, const float* s, float* scale, const float* g, float eps, size_t ns, size_t num)
+        {
+            // Compute sum of squares
+            for (auto n : grid_stride_range_y(0, ns))
+            {
+                auto p = s + n * num;
+                float sum_squares = 0;
+                for (auto i : grid_stride_range(0, num))
+                {
+                    sum_squares += p[i] * p[i];
+                }
+                warp_reduce_atomic_add(scale[n], sum_squares / num);
+            }
+            __syncthreads();
+
+            // Compute RMS inverse
+            for (auto n : grid_stride_range_y(0, ns))
+            {
+                for (auto i : grid_stride_range(0, 1))
+                {
+                    scale[n] = 1.0f / std::sqrt(scale[n] + eps);
+                }
+            }
+            __syncthreads();
+
+            for (auto n : grid_stride_range_y(0, ns))
+            {
+                for (auto i : grid_stride_range(0, num))
+                {
+                    const float val = s[n * num + i] * scale[n];
+                    out[n * num + i] = val * g[i];
+                }
+            }
+        }
+
+        __global__ void _cuda_rms_normalize_gradient(float* out, float* gg, const float* s, const float* gi, const float* scale, const float* g, float* dscale, float eps, size_t ns, size_t num)
+        {
+            for (auto n : grid_stride_range_y(0, ns))
+            {
+                float temp_dscale = 0;
+                for (auto i : grid_stride_range(0, num))
+                {
+                    auto idx = n * num + i;
+                    const float x_hat = s[idx] * scale[n];
+                    gg[i] += gi[idx] * x_hat;
```

Comment:
This needs to use a `warp_reduce_atomic_add()`.  Can't have multiple threads adding to the same output.

Reply:
Regarding the concurrent write issue, my initial code already included a mechanism to protect this operation:

`atomicAdd(&gamma_grad[c], sum_grad_x_hat_for_gamma);`

Reply:
The updated code here after:

```
        __global__ void rms_normalize_gradient_kernel(
            const float* gradient_input,
            const float* src,
            const float* gamma,
            const float* scale,
            float* src_grad,
            float* gamma_grad,
            float* dscale,
            const long num_samples,
            const long num_channels,
            const long spatial_dim
        )
        {
            const long n = blockIdx.x;
            const long c = threadIdx.x;

            if (n < num_samples && c < num_channels)
            {
                float sum_grad_x_hat = 0;
                float sum_grad_x_hat_x = 0;
                float sum_grad_x_hat_for_gamma = 0;
                float rms_inv = scale[n * num_channels + c];

                for (long s = 0; s < spatial_dim; ++s)
                {
                    long idx = (n * num_channels + c) * spatial_dim + s;
                    float x = src[idx];
                    float grad = gradient_input[idx];
                    float x_hat = x * rms_inv;

                    sum_grad_x_hat += grad * gamma[c];
                    sum_grad_x_hat_x += grad * gamma[c] * x_hat * x;
                    sum_grad_x_hat_for_gamma += grad * x_hat;
                }

                float dscale_val = -0.5f * rms_inv * rms_inv * rms_inv * sum_grad_x_hat_x;
                dscale[n * num_channels + c] = dscale_val;

                atomicAdd(&gamma_grad[c], sum_grad_x_hat_for_gamma);

                for (long s = 0; s < spatial_dim; ++s)
                {
                    long idx = (n * num_channels + c) * spatial_dim + s;
                    float x = src[idx];
                    float grad = gradient_input[idx];
                    float x_hat = x * rms_inv;

                    src_grad[idx] = gamma[c] * rms_inv * (grad - (sum_grad_x_hat * x_hat + dscale_val * x * 2) / spatial_dim);
                }
            }
        }
```
Looking at it this way, I don't see much of a concurrency problem...

Reply:
Yeah that doesn't look to have any data races in it.   Totally use a warp reduction with a grid stride loop though.  Those are super fast.  Since the warp reduction stuff is extremely fast and using a nested grid stride loop is way faster than a regular `for` loop since that uses cuda threads.

I.e. this pattern:
```
                        // loop over some 2D thing and sum and store things into out.
                        for (auto i : grid_stride_range_y(0, nr))
                        {
                            float temp = 0;
                            for (auto j : grid_stride_range(0, nc))
                                temp += whatever[i*nc+j];

                            // store the sum into out[i]
                            warp_reduce_atomic_add(out[i], temp);
                        }
```
It's wildly fast.  More about it at https://developer.nvidia.com/blog/faster-parallel-reductions-kepler/ which is what that warp_reduce_atomic_add() is doing.

Reply:
Yes, I've completely reworked the code to fix the competition problems and use the reduction mechanisms.